src/main/java/com/ibm/darpc/DaRPCMemPoolImpl.java

```diff
@@ -0,0 +1,159 @@
+	ByteBuffer allocateHugePageBuffer() throws IOException {
+		if (hugePagePath == null) {
+			System.out.println("Hugepage path must be set");
+			throw new IOException("Hugepage path must be set");
+		}
+
+		hugePageFile = hugePagePath + "/darpcmempoolimpl.mem";
+		RandomAccessFile randomFile = null;
+		try {
+			randomFile = new RandomAccessFile(hugePageFile, "rw");
+		} catch (FileNotFoundException e) {
+			System.out.println("Path " + hugePageFile + " to huge page directory not found.");
+			throw e;
+		}
+		try {
+			randomFile.setLength(allocationSize + alignmentSize);
+		} catch (IOException e) {
+			System.out.println("Coult not set allocation length of mapped random access file on huge page directory.");
+			System.out.println("allocaiton size = " + allocationSize + " , alignment size =  " + alignmentSize);
+			System.out.println("allocation size and alignment must be a multiple of the hugepage size.");
+			randomFile.close();
+			throw e;
+		}
+		FileChannel channel = randomFile.getChannel();
+		MappedByteBuffer mappedBuffer = null;
+		try {
+			mappedBuffer = channel.map(MapMode.READ_WRITE, 0,
+					allocationSize + alignmentSize);
+		} catch (IOException e) {
+			System.out.println("Could not map the huge page file on path " + hugePageFile);
+			randomFile.close();
+			throw e;
+		}
+		randomFile.close();
+
+		long rawBufferAddress = MemoryUtils.getAddress(mappedBuffer);
+		if (alignmentSize > 0) {
+			long alignmentOffset = rawBufferAddress % alignmentSize;
+			if (alignmentOffset != 0) {
+				mappedBuffer.position(alignmentSize - (int)alignmentOffset);
+			}
+		}
+		ByteBuffer b = mappedBuffer.slice();
+		return (b);
+	}
+
+
+	@Override
+	public void close() throws IOException {
+		synchronized(this) {
+			try {
+				mr.deregMr().execute().free();
+			} catch (IOException e) {
+				System.out.println("Could not unregister memory region.");
+				e.printStackTrace();
+			}
+			File f = new File(hugePageFile);
+			f.delete();
+		}
+	}
+
+	@Override
+	public ByteBuffer getBuffer(RdmaEndpoint endpoint, int size) throws IOException {
+		ByteBuffer r = null;
+
+		synchronized(this) {
+			if (pd == null) {
+				pd = endpoint.getPd();
+			} else if (!pd.equals(endpoint.getPd())) {
+				throw new IOException("No support for more than one PD");
+			}
+			if (mr == null) {
+				byteBuffer = allocateHugePageBuffer();
+				mr = pd.regMr(byteBuffer, access).execute().free().getMr();
+			}
+
+			if (freeList == null) {
+				offset = size;
+				freeList = new LinkedList<ByteBuffer>();
+				int i = 0;
+				while ((i * offset + offset) < byteBuffer.capacity()) {
+					byteBuffer.position(i * offset);
+					byteBuffer.limit(i * offset + offset);
+					ByteBuffer b = byteBuffer.slice();
+					freeList.addLast(b);
+					i++;
+				}
+			}
+			else
+			{
+				if (size != offset) {
+					throw new IOException("Requested size does not match block size managed by memory pool.");
+				}
+			}
+			r = freeList.removeFirst();
```

Comment:
This throws a NoSuchElementException if the freeList is empty: https://docs.oracle.com/javase/7/docs/api/java/util/LinkedList.html#removeFirst()
This is not obvious from the interface description.

Reply:
Yes, I agree. Changed interface showing that it throws a NoSuchElementException.